File: app/src/main/java/com/u1fukui/bbs/ui/main/home/HomeFragment.java
package com.u1fukui.bbs.ui.main.home;

import android.databinding.ObservableList;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.u1fukui.bbs.databinding.FragmentHomeBinding;
import com.u1fukui.bbs.model.Category;

import javax.inject.Inject;

import dagger.android.support.DaggerFragment;

public class HomeFragment extends DaggerFragment {

    public static final String TAG = HomeFragment.class.getSimpleName();

    @Inject
    HomeViewModel viewModel;

    private FragmentHomeBinding binding;

    public static HomeFragment newInstance() {
        return new HomeFragment();
    }

    public HomeFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentHomeBinding.inflate(inflater, container, false);
        binding.setViewModel(viewModel);
        initViews();

        viewModel.start();

        return binding.getRoot();
    }

    private void initViews() {
        binding.viewPager.setAdapter(new HomePagerAdapter(getChildFragmentManager(), viewModel.getCategoryList()));
        binding.tabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);
        binding.tabLayout.setupWithViewPager(binding.viewPager);
    }

    @Override
    public void onDestroyView() {
        binding.unbind();
        super.onDestroyView();
    }

    private static class HomePagerAdapter extends FragmentPagerAdapter {

        private final ObservableList<Category> categoryList;

        public HomePagerAdapter(FragmentManager manager, ObservableList<Category> categoryList) {
            super(manager);
            this.categoryList = categoryList;
            this.categoryList.addOnListChangedCallback(new ObservableList.OnListChangedCallback<ObservableList<Category>>() {
                @Override
                public void onChanged(ObservableList<Category> sender) {
                    notifyDataSetChanged();
                }

                @Override
                public void onItemRangeChanged(ObservableList<Category> sender, int positionStart, int itemCount) {
                    notifyDataSetChanged();
                }

                @Override
                public void onItemRangeInserted(ObservableList<Category> sender, int positionStart, int itemCount) {
                    notifyDataSetChanged();
                }

                @Override
                public void onItemRangeMoved(ObservableList<Category> sender, int fromPosition, int toPosition, int itemCount) {
                    notifyDataSetChanged();
                }

                @Override
                public void onItemRangeRemoved(ObservableList<Category> sender, int positionStart, int itemCount) {
                    notifyDataSetChanged();
                }
            });
        }

        @Override
        public Fragment getItem(int position) {
            return CategoryThreadListFragment.newInstance(categoryList.get(position));
        }

        @Override
        public int getCount() {
            return categoryList.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return categoryList.get(position).getName();
        }
    }
}

File: app/src/main/java/com/u1fukui/bbs/ui/main/home/HomeFragment.kt
package com.u1fukui.bbs.ui.main.home

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.ObservableList
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.ViewModelProviders
import com.google.android.material.tabs.TabLayout
import com.u1fukui.bbs.databinding.FragmentHomeBinding
import com.u1fukui.bbs.model.Category
import com.u1fukui.bbs.repository.CategoryListRepository
import com.u1fukui.bbs.ui.main.MainActivity
import com.u1fukui.bbs.ui.main.MainNavigator
import dagger.android.support.DaggerFragment

class HomeFragment : DaggerFragment() {

    private lateinit var binding: FragmentHomeBinding

    private val viewModel: HomeViewModel by lazy {
        ViewModelProviders
                .of(this, HomeViewModel.Factory(
                        CategoryListRepository(),
                        MainNavigator(activity as MainActivity)
                ))
                .get(HomeViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentHomeBinding.inflate(inflater, container, false)
        binding.viewModel = viewModel
        initViews()

        viewModel.start()

        return binding.root
    }

    private fun initViews() {
        binding.apply {
            viewPager.adapter = HomePagerAdapter(childFragmentManager, viewModel!!.categoryList)
            tabLayout.tabMode = TabLayout.MODE_SCROLLABLE
            tabLayout.setupWithViewPager(binding.viewPager)
        }
    }

    override fun onDestroyView() {
        binding.unbind()
        super.onDestroyView()
    }

    private class HomePagerAdapter(
        manager: FragmentManager,
        private val categoryList: ObservableList<Category>
    ) : FragmentPagerAdapter(manager) {

        init {
            this.categoryList.addOnListChangedCallback(object : ObservableList.OnListChangedCallback<ObservableList<Category>>() {
                override fun onChanged(sender: ObservableList<Category>) {
                    notifyDataSetChanged()
                }

                override fun onItemRangeChanged(sender: ObservableList<Category>, positionStart: Int, itemCount: Int) {
                    notifyDataSetChanged()
                }

                override fun onItemRangeInserted(sender: ObservableList<Category>, positionStart: Int, itemCount: Int) {
                    notifyDataSetChanged()
                }

                override fun onItemRangeMoved(sender: ObservableList<Category>, fromPosition: Int, toPosition: Int, itemCount: Int) {
                    notifyDataSetChanged()
                }

                override fun onItemRangeRemoved(sender: ObservableList<Category>, positionStart: Int, itemCount: Int) {
                    notifyDataSetChanged()
                }
            })
        }

        override fun getItem(position: Int) = CategoryThreadListFragment.newInstance(categoryList[position])

        override fun getCount() = categoryList.size

        override fun getPageTitle(position: Int) = categoryList[position].name
    }

    companion object {

        @JvmStatic
        val TAG = HomeFragment::class.java.simpleName

        @JvmStatic
        fun newInstance() = HomeFragment()
    }
}
